solr/prometheus-exporter/src/java/org/apache/solr/prometheus/collector/MetricSamples.java

```diff
@@ -47,7 +65,8 @@ public void addSampleIfMetricExists(
       return;
     }
 
-    if (!sampleFamily.samples.contains(sample)) {
+    if (!sampleMetricsCache.contains(sample.toString())) {
+      sampleMetricsCache.add(sample.toString());
```

Comment:
Thanks @mlbiscoc for explaining the issue in the PR description, I'm new to this part of the code base and the explanation really helped!

So here the `sampleFamily.samples.contains` is inefficient for a large list and the idea is that a `HashSet.contains` would be more efficient.

It seems that `sample.toString()` doesn't include the `exemplar` member -- https://github.com/prometheus/client_java/blob/parent-0.16.0/simpleclient/src/main/java/io/prometheus/client/Collector.java#L256-L292 -- i.e. comparison of `sample` vs. comparison via `sample.toString()` could theoretically differ? Hmm.

Reply:
No problem! I am also still learning this code base as well but please ask anything if it needs clarification.

Exactly, this PR is to address the fact that we are searching in `O(n)` but we can change that to `O(1)`. Testing locally, I found significant performance increase.

As for the `.toString` not including `exemplar`, I felt it was a non-issue because when we scrape the metrics from Solr, the prometheus exporter creates the sample object, but with `timestamp and exemplar = null`. So there will never be a string with a `timestamp` or `exemplar`. You can see here -> https://github.com/apache/solr/blob/7a31a1e23f6034c14639d7a208c5aaec64608cfd/solr/prometheus-exporter/src/java/org/apache/solr/prometheus/scraper/SolrScraper.java#L205

Reply:
Again, you could use a sampleMetricCache.add(sample.toString()) and look at the response boolean to know if it was added (wasn't already present).  Is more efficient and moreover avoids a second sample.toString() call easily.

Why is sampleMetricsCache using the toString of the Sample at all; why isn't it simply `Set<Sample>`?  Also; sampleMetricsCache isn't a cache; it's not the ideal name/word for it.  I would call it "seenSamples".